File: code/decorator/decorator.java
interface Image {
    // Returning string for the sake of a demo
    public String getImage();
}

class PatternedImage implements Image {
    private final String description;

    public PatternedImage(String desc) {
        description = desc;
    }

    @Override
    public String getImage() {
        return description;
    }
}

abstract class ImageDecorator implements Image {
    protected Image toBeDecorated;

    public ImageDecorator(Image toBeDecorated) {
        this.toBeDecorated = toBeDecorated;
    }

    @Override
    public String getImage() {
        return toBeDecorated.getImage();
    }
}

class ContrastFilter extends ImageDecorator {
    protected double factor;

    public ContrastFilter(double factor, Image toBeDecorated) {
        super(toBeDecorated);
        this.factor = factor;
    }

    @Override
    public String getImage() {
        return super.getImage() + " + Contrast factor " + factor;
    }
}

class MonochromeFilter extends ImageDecorator {
    public MonochromeFilter(Image toBeDecorated) {
        super(toBeDecorated);
    }

    @Override
    public String getImage() {
        return super.getImage() + " + Monochrome conversion";
    }
}

class EdgeDetectionFilter extends ImageDecorator {
    public EdgeDetectionFilter(Image toBeDecorated) {
        super(toBeDecorated);
    }

    @Override
    public String getImage() {
        return super.getImage() + " + Edge detection";
    }
}

class Main {
    public static void main(String[] args) {
        Image img = new PatternedImage("Mosaic Image");

        Image transformedImage = new EdgeDetectionFilter(
            // new MonochromeFilter(
                new ContrastFilter(
                    2.5,
                    img
                )
            // )
        );

        System.out.println(transformedImage.getImage());
    }
}

File: code/decorator/decorator.hs
type Image = String

contrastFilter :: Double -> Image -> Image
contrastFilter factor img = img ++ " + Contrast factor " ++ show(factor)

monochromeFilter :: Image -> Image
monochromeFilter img = img ++ " + Monochrome conversion"

edgeDetectionFilter :: Image -> Image
edgeDetectionFilter img = img ++ " + Edge detection"

combineFilters :: [(Image -> Image)] -> (Image -> Image)
combineFilters fs = foldl (flip (.)) id fs

main :: IO()
main = do
  -- Static version
  putStrLn $ (edgeDetectionFilter . (contrastFilter 2.5) . monochromeFilter) "Mosaic image"
  
  -- dynamic version
  putStrLn $ filter "Mosaic Image"
  where
    filter = combineFilters [monochromeFilter, (contrastFilter 2.5), edgeDetectionFilter]